src/cwhy/cwhy.py

```diff
@@ -186,95 +188,137 @@ def evaluate_text_prompt(args, prompt, wrap=True, **kwargs):
 
 class explain_context:
     def __init__(self, args, diagnostic):
-        diagnostic_lines = diagnostic.splitlines()
+        self.args = args
+        self.encoding = tiktoken.encoding_for_model(args["llm"])
+        self.diagnostic_lines = diagnostic.splitlines()
 
-        self.code_locations = {}
-
-        # Don't send more than this many code locations.
-        # This is just to prevent overwhelming OpenAI.
-        max_code_locations = args["max_context"]
+        # We group by source file, then keep line numbers ordered.
+        self.code_locations = collections.defaultdict(collections.OrderedDict)
 
         # Go through the diagnostic and build up a list of code locations.
-        line = 0
-        while line < len(diagnostic_lines):
+        for line in self.diagnostic_lines:
             # This pattern works for some C++ compilers (GCC, Clang) and Rust.
-            match = re.match(
-                r"([a-zA-Z0-9./][^:->]+):([0-9]+):([0-9]+)", diagnostic_lines[line]
-            )
+            match = re.match(r"([a-zA-Z0-9./][^:->]+):([0-9]+):([0-9]+)", line)
 
             if not match:
                 # This pattern works for javac.
-                match = re.match(
-                    r"([a-zA-Z0-9./][^:->]+):([0-9]+):", diagnostic_lines[line]
-                )
+                match = re.match(r"([a-zA-Z0-9./][^:->]+):([0-9]+):", line)
 
             if not match:
                 # This pattern works for Python, filtering out non-files (e.g., <string>).
-                match = re.match(
-                    r'\s*File "(.*?)", line (\d+), in ([^\<].*)', diagnostic_lines[line]
-                )
-
-            line += 1
+                match = re.match(r'\s*File "(.*?)", line (\d+), in ([^\<].*)', line)
 
             if not match:
                 continue
 
-            if max_code_locations == 0:
-                # We've found the end of the last "frame", and we don't have room
-                # for anymore, so we're done.
-                break
-
             file_name = match.group(1).lstrip()
             line_number = int(match.group(2))
 
             try:
-                (abridged_code, line_start, line_end) = read_lines(
-                    file_name, line_number - 7, line_number + 2
+                (abridged_code, line_start) = read_lines(
+                    file_name, line_number - 7, line_number + 3
```

Comment:
Should we consider making these a parameter?

Reply:
I will do this separately.